### PR TITLE
Issue 3437 new segment stringinterning

### LINKED ISSUE #3437: Fail to select new segment using segid just added to the universe
## Expected behavior ##

I was trying to add a new segment to the universe and select that segment from the universe. When I select the atoms assigned to the new segment by the previous step, python returns an index error.  

## Actual behavior ##

```
Traceback (most recent call last):
 File "<stdin>", line 1, in <module>
 File "~/python3.6/site-packages/MDAnalysis/core/universe.py", line 677, in select_
atoms
   return self.atoms.select_atoms(*args, **kwargs)
 File "~/python3.6/site-packages/MDAnalysis/core/groups.py", line 2916, in select_a
toms
   selections[0].apply(self))
 File "~/python3.6/site-packages/MDAnalysis/core/selection.py", line 173, in apply
   return func(self, group)
 File "~/python3.6/site-packages/MDAnalysis/core/selection.py", line 549, in apply
   nmidx = nmattr.nmidx[getattr(group, self.level)]
IndexError: index 2 is out of bounds for axis 0 with size 2
```

## Code to reproduce the behavior ##

```
>>> import MDAnalysis as mda
>>> from MDAnalysis.tests import datafiles as data

>>> u = mda.Universe(data.RNA_PSF, data.RNA_PDB)
>>> u.segments.segids
array(['POT', 'RNAA', 'SOLV'], dtype=object)

>>> atomP = u.add_Segment(segid='P')
>>> P_atoms = u.select_atoms("name P")
>>> P_atoms.residues.segments = atomP

>>> u.segments.segids
array(['POT', 'RNAA', 'SOLV', 'P'], dtype=object)

>>> u.select_atoms("segid P”)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-22-201611c1d061> in <module>
----> 1 u.select_atoms("segid P”)
...
...
IndexError: index 3 is out of bounds for axis 0 with size 3
```
## Current version of MDAnalysis ##

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)
- 1.1.1
- Which version of Python (`python -V`)?
- Python 3.6.13
- Which operating system?
- Linux Opensuse

### PR DESCRIPTION
**Pull request by richardjgowers**
Fixes #3437

Changes made in this Pull Request:

The string interning we're using for string TopologyAttrs uses a more complex system than just a numpy array of values.  `Topology.add_Residue` and `Topology.add_Segment` previously just assumed a numpy array of values, so blindly "appended" values to this without respecting fancy data structures.

The fix is to add `_add_new` to each TopologyAttr which is a respectful way of appending to the TopologyAttr.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Review by richardjgowers — COMMENTED** [review]
@MDAnalysis/coredevs @p-j-smith   Can anyone review this please?

File: package/MDAnalysis/core/topologyattrs.py [hunk 1]
```diff
@@ -1990,6 +2043,18 @@ def set_segments(self, sg, values):
         raise _wronglevel_error(self, sg)
 
 
+# woe betide anyone who switches this inheritance order
+# Mixin needs to be first (L to R) to get correct __init__ and set_atoms
+class _ResidueStringAttr(_StringInternerMixin, ResidueAttr):
```
Comment (lilyminium) [inline package/MDAnalysis/core/topologyattrs.py]: Could this please not be private, so users can easily create new string attributes?

That or we need a new ProtoStringSelection class that does not rely on the these structures

File: testsuite/MDAnalysisTests/core/test_topologyattrs.py [hunk 2]
```diff
@@ -585,3 +585,37 @@ def test_deprecate_bfactors_set_atom(self, universe):
...
+    @pytest.mark.parametrize('newname', ['SegA', 'SegB'])
+    def test_add_segment(self, universe, newname):
+        newseg = universe.add_Segment(segid=newname)
+
+        assert newseg.segid == newname
+
+        rg = universe.residues[0]
+        rg.segment = newseg
+
+        assert rg.atoms[0].segid == newname
```
Comment (lilyminium) [inline testsuite/MDAnalysisTests/core/test_topologyattrs.py:622]: could you please add a test for the issue in #3437  with the selection language?

File: package/MDAnalysis/core/topologyattrs.py [hunk 3]
```diff
@@ -712,6 +754,17 @@ def set_atoms(self, ag, values):
...
+# woe betide anyone who switches this inheritance order
+# Mixin needs to be first (L to R) to get correct __init__ and set_atoms
+class _AtomStringAttr(_StringInternerMixin, AtomAttr):
+    @_check_length
+    def set_atoms(self, ag, values):
+        return self._set_X(ag, values)
+
+    @staticmethod
+    def _gen_initial_values(na, nr, ns):
+        return np.full('', na, dtype=object)
```
Comment (lilyminium) [inline package/MDAnalysis/core/topologyattrs.py]: Got this the wrong way around (and in other np.fulls)

``` numpy.full(shape, fill_value, dtype=None, order='C', *, like=None)```

**Review by lilyminium — CHANGES_REQUESTED** [review]

**Comment by richardjgowers** [comment]
thanks @lilyminium

**Comment by orbeckst** [comment]
I like the reduction in code size :-). 

I also like the use of the phrase "woe betide `<OBJECT>`". 

`echo 🥕| sed 's/orange/white/'` for the win (can't find a turnip or parsnip emoji)

**Review by lilyminium — APPROVED** [review]

**Comment by lilyminium** [comment]
LGTM, just fixed the merge conflict, good to go if tests remain green. Thank you :-)

**Pull request merged**